### PR TITLE
mgr/nfs: Don't crash ceph-mgr if NFS clusters are unavailable

### PR DESCRIPTION
**Pull request by pponnuvel**
Fixes https://tracker.ceph.com/issues/56246.
Fixes https://tracker.ceph.com/issues/59438.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by pponnuvel** [comment]
jenkins test dashboard cephadm

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by nizamial09** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
jenkins retest this please

**Comment by pponnuvel** [comment]
Looks like checking out the branch doesn't work:
```
 > git fetch --tags --force --progress --depth=1 -- https://github.com/ceph/ceph.git +refs/heads/*:refs/remotes/origin/* # timeout=20
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --tags --force --progress --depth=1 -- https://github.com/ceph/ceph.git +refs/heads/*:refs/remotes/origin/*" returned status code 128:
stdout: 
stderr: fatal: unable to access 'https://github.com/ceph/ceph.git/': Could not resolve host: github.com

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2842)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2185)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:635)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:871)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:170)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:377)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:125)
	at java.base/java.lang.Thread.run(Thread.java:829)
	Suppressed: hudson.remoting.Channel$CallSiteStackTrace: Remote call to JNLP4-connect connection from 8.43.84.3/8.43.84.3:53523
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1787)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1003)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:153)
		at jdk.internal.reflect.GeneratedMethodAccessor679.invoke(Unknown Source)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
		at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:138)
		at com.sun.proxy.$Proxy189.execute(Unknown Source)
		at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1222)
		at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1305)
		at hudson.scm.SCM.checkout(SCM.java:540)
		at hudson.model.AbstractProject.checkout(AbstractProject.java:1248)
		at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:649)
		at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:85)
		at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:521)
		at hudson.model.Run.execute(Run.java:1895)
		at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:44)
		at hudson.model.ResourceController.execute(ResourceController.java:101)
		at hudson.model.Executor.run(Executor.java:442)
ERROR: Error cloning remote repo 'origin'
```

**Comment by pponnuvel** [comment]
I tested locally and the dashboard works as expected. I'll wait for the CI issues to be resolved and tests to pass before marking this ready for review.

**Comment by nizamial09** [comment]
jenkins retest this please

**Comment by pponnuvel** [comment]
jenkins test dashboard cephadm

**Comment by pponnuvel** [comment]
jenkins test dashboard

**Comment by pponnuvel** [comment]
@nizamial09  Looks like the cypress tests are failing. It seems the Applitools results aren't available to everyone. Can you please take a look?

**Comment by nizamial09** [comment]
> @nizamial09 Looks like the cypress tests are failing. It seems the Applitools results aren't available to everyone. Can you please take a look?

i think you can ignore that error. We are fixing that separately.

**Comment by nizamial09** [comment]
@avanthakkar can you have a look and review this PR?

**Review by nizamial09 — COMMENTED** [review]
btw, thank you for the contribution. Just letting you know that we have a guidelines for contribution which [you can find here](https://github.com/ceph/ceph/blob/main/SubmittingPatches-backports.rst).

To summarize, you need to re-write the title as `mgr/nfs: don't crash ...` since you are mostly modifying the nfs module code. Also you need to link the tracker ticket in the commit as `Fixes: <url-of-tracker>`. You need to create a tracker [here](https://tracker.ceph.com/), incase its not already present.

**Comment by pponnuvel** [comment]
> btw, thank you for the contribution. Just letting you know that we have a guidelines for contribution which [you can find here](https://github.com/ceph/ceph/blob/main/SubmittingPatches-backports.rst).
> 
> To summarize, you need to re-write the title as `mgr/nfs: don't crash ...` since you are mostly modifying the nfs module code. Also you need to link the tracker ticket in the commit as `Fixes: <url-of-tracker>`. You need to create a tracker [here](https://tracker.ceph.com/), incase its not already present.

Fixed the commit message and tracker links. Thanks.

**Comment by pponnuvel** [comment]
@avanthakkar @nizamial09 Can you please take a look at this? Thanks.

**Review by avanthakkar — CHANGES_REQUESTED** [review]
Firstly, if there are any issues with the Dashboard crashing, my preference would be to address them within the dashboard code rather than making changes to the NFS module. The API should still be able to manage exceptions and return empty responses if no orchestrator is available. It's worth noting that altering the NFS module code also impacts the behavior of CLI commands. From what I understand, the management of the NFS cluster relies on the orchestrator. For instance, if you examine the `nfs cluster ls` command, it utilizes:

```
completion = mgr.describe_service(service_type='nfs')
```

Where `describe_service` is defined within the cephadm module. Therefore, my suggestion would be to modify the response of the `/nfs-ganesha/cluster'` API to handle exceptions and return empty results if no orchestrator is present.

Reference: [GitHub Link](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/controllers/nfs.py#L86)

**Comment by pponnuvel** [comment]
> Firstly, if there are any issues with the Dashboard crashing, my preference would be to address them within the dashboard code rather than making changes to the NFS module. The API should still be able to manage exceptions and return empty responses if no orchestrator is available. It's worth noting that altering the NFS module code also impacts the behavior of CLI commands. From what I understand, the management of the NFS cluster relies on the orchestrator. For instance, if you examine the `nfs cluster ls` command, it utilizes:
> 
> ```
> completion = mgr.describe_service(service_type='nfs')
> ```
> 
> Where `describe_service` is defined within the cephadm module. Therefore, my suggestion would be to modify the response of the `/nfs-ganesha/cluster'` API to handle exceptions and return empty results if no orchestrator is present.
> 
> Reference: [GitHub Link](https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/controllers/nfs.py#L86)

Thanks, @avanthakkar. To be clear, it crashes the ceph-mgr and that results in ceph-mgr fail over.

I think your suggestion makes sense and I'll modify to handle this within dashboard itself (without changing nfs module). Thanks.

File: src/pybind/mgr/nfs/utils.py
Comment (avanthakkar) [inline src/pybind/mgr/nfs/utils.py:71]: Functionally it looks good to me! Can you attach a screenshot or a screencast of how Dashboard behaves now with changes? Thank you!

Comment (pponnuvel) [reply src/pybind/mgr/nfs/utils.py:71]: Current behaviour on clicking 'NFS' tab:
ceph-dashboard-nfs-current-behaviour

New behaviour on clicking 'NFS' tab:
ceph-dashboard-nfs-new-behaviour1

On clicking 'create' (above):
ceph-dashboard-nfs-new-behaviour2

Then on clicking 'add a new NFS Service':
ceph-dashboard-nfs-new-behaviour3


===

Current log excerpts from ceph-mgr:
```
2024-04-02T17:29:31.074+0000 7b26127fe640 -1 Remote method threw exception: Traceback (most recent call last): 
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/module.py", line 189, in cluster_ls
    return available_clusters(self)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/utils.py", line 70, in available_clusters
    completion = mgr.describe_service(service_type='nfs')
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1689, in inner
    completion = self._oremote(method_name, args, kwargs)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1756, in _oremote
    raise NoOrchestrator()
orchestrator._interface.NoOrchestrator: No orchestrator configured (try `ceph orch set backend`)

2024-04-02T17:29:31.078+0000 7b26127fe640  0 [dashboard ERROR controllers.nfs] Remote method threw exception: Traceback (most recent call last): 
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/module.py", line 189, in cluster_ls
    return available_clusters(self)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/utils.py", line 70, in available_clusters
    completion = mgr.describe_service(service_type='nfs')
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1689, in inner
    completion = self._oremote(method_name, args, kwargs)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1756, in _oremote
    raise NoOrchestrator()
orchestrator._interface.NoOrchestrator: No orchestrator configured (try `ceph orch set backend`)
Traceback (most recent call last): 
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/dashboard/controllers/nfs.py", line 273, in status 
    mgr.remote('nfs', 'cluster_ls')
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/mgr_module.py", line 2280, in remote 
    return self._ceph_dispatch_remote(module_name, method_name,
RuntimeError: Remote method threw exception: Traceback (most recent call last): 
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/module.py", line 189, in cluster_ls
    return available_clusters(self)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/utils.py", line 70, in available_clusters
    completion = mgr.describe_service(service_type='nfs')
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1689, in inner
    completion = self._oremote(method_name, args, kwargs)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1756, in _oremote
    raise NoOrchestrator()
orchestrator._interface.NoOrchestrator: No orchestrator configured (try `ceph orch set backend`)

2024-04-02T17:29:31.082+0000 7b26127fe640  0 [dashboard INFO request] [::ffff:127.0.0.1:37338] [GET] [200] [0.059s] [admin] [360.0B] /ui-api/nfs-ganesha/status
```

New log excerpts from ceph-mgr:
```
2024-04-02T17:37:40.937+0000 7dfd1a3fa640 20 mgr ~Gil Destroying new thread state 0x5c74f9c619e0
2024-04-02T17:37:40.945+0000 7dfd1a3fa640  0 [nfs ERROR nfs.utils] No orchestrator configured
Traceback (most recent call last):
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/nfs/utils.py", line 72, in available_clusters
    completion = mgr.describe_service(service_type='nfs')
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1689, in inner 
    completion = self._oremote(method_name, args, kwargs)
  File "/home/ponnuvel/cases/ceph/src/pybind/mgr/orchestrator/_interface.py", line 1756, in _oremote
    raise NoOrchestrator()
orchestrator._interface.NoOrchestrator: No orchestrator configured (try `ceph orch set backend`)
```

Comment (pponnuvel) [reply src/pybind/mgr/nfs/utils.py:71]: @avanthakkar I've uploaded screenshots as well the relevant log from ceph-mgr here. Please take a look. Thanks!

**Review by avanthakkar — APPROVED** [review]
LGTM! Thanks @pponnuvel

**Comment by pponnuvel** [comment]
jenkins test make check arm64

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
jenkins test make check arm64

**Comment by avanthakkar** [comment]
> @avanthakkar @nizamial09 Good to merge? Please let me know if anything I can do. Thanks.

Looks good to me for merging. I noticed you closed the PR. Is there a specific reason for that?

**Comment by pponnuvel** [comment]
> > @avanthakkar @nizamial09 Good to merge? Please let me know if anything I can do. Thanks.
> 
> Looks good to me for merging. I noticed you closed the PR. Is there a specific reason for that?

Ah, I hadn't realised deleting repo would also close PR automatically (I deleted & re-fork ceph repo).
I've created this PR again: https://github.com/ceph/ceph/pull/56876

**Comment by avanthakkar** [comment]
> > > @avanthakkar @nizamial09 Good to merge? Please let me know if anything I can do. Thanks.
> > 
> > 
> > Looks good to me for merging. I noticed you closed the PR. Is there a specific reason for that?
> 
> Ah, I hadn't realised deleting repo would also close PR automatically (I deleted & re-fork ceph repo). I've created this PR again: #56876

Ah I see, you can delete branch once the PR is merged

**Comment by avanthakkar** [comment]
Reopening PR as it was closed by mistake

**Pull request closed without merging**
